### ISSUE TITLE
Is there a way to redirect to previous page after redirect_route?

### ISSUE THREAD

**Issue by rainje** [knpuniversity/oauth2-client-bundle#73]
Hi,

I've been using this for months now with no issues but I'd like to enhance it by redirecting back the user connecting to its previously visiting page.

I'm using both Facebook and Google clients with the Guard implementation and both my authenticators have onAuthenticationSuccess set to redirect to my homepage. Which sucks.

Here's my FacebookAuthenticator.php
```
<?php

namespace AppBundle\Security;

use AppBundle\Entity\User;
use KnpU\OAuth2ClientBundle\Security\Exception\FinishRegistrationException;
use KnpU\OAuth2ClientBundle\Security\Helper\FinishRegistrationBehavior;
use Doctrine\ORM\EntityManager;
use KnpU\OAuth2ClientBundle\Client\ClientRegistry;
use KnpU\OAuth2ClientBundle\Security\Helper\PreviousUrlHelper;
use KnpU\OAuth2ClientBundle\Security\Helper\SaveAuthFailureMessage;
use Symfony\Component\DependencyInjection\ContainerInterface;
use League\OAuth2\Client\Token\AccessToken;
use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
use League\OAuth2\Client\Provider\Facebook;
use League\OAuth2\Client\Provider\FacebookUser;
use Symfony\Component\HttpFoundation\RedirectResponse;
use Symfony\Component\Routing\RouterInterface;
use KnpU\OAuth2ClientBundle\Security\Authenticator\SocialAuthenticator;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
use Symfony\Component\Security\Core\Exception\AuthenticationException;
use Symfony\Component\Security\Core\User\UserInterface;
use Symfony\Component\Security\Core\User\UserProviderInterface;
use KnpU\OAuth2ClientBundle\Client\Provider\FacebookClient;
use Symfony\Component\Security\Core\Encoder\UserPasswordEncoder;

class FacebookAuthenticator extends SocialAuthenticator
{
    private $clientRegistry;
    private $em;
    private $router;
    private $encoder;
    private $locales;

    public function __construct(ClientRegistry $clientRegistry, EntityManager $em, RouterInterface $router, UserPasswordEncoder $encoder, $locales)
    {
        $this->clientRegistry = $clientRegistry;
        $this->em = $em;
        $this->router = $router;
        $this->encoder = $encoder;
        $this->locales = $locales;
    }

    public function getCredentials(Request $request)
    {
        foreach ($this->locales as $locale) {
            if ($request->getPathInfo() == '/' . $locale . '/connect/facebook/check') {
                return $this->fetchAccessToken($this->getFacebookClient());
            }
        }
        return;
    }

    public function getUser($credentials, UserProviderInterface $userProvider)
    {
        /** @var FacebookUser $facebookUser */
        $facebookUser = $this->getFacebookClient()
            ->fetchUserFromToken($credentials);
        $email = $facebookUser->getEmail();

        //do stuff with the user (add to db... or logging in if exists)
        return $user;
    }

    /**
     * @return FacebookClient
     */
    private function getFacebookClient()
    {
        return $this->clientRegistry
            // "facebook_main" is the key used in config.yml
            ->getClient('facebook');
    }

    public function checkCredentials($credentials, UserInterface $user)
    {
        // check credentials - e.g. make sure the password is valid
        // no credential check is needed in this case

        // return true to cause authentication success
        return true;
    }

    /**
     * @param Request $request
     * @param AuthenticationException $exception
     *
     * @return Response|null
     */
    public function onAuthenticationFailure(Request $request, AuthenticationException $exception)
    {
        $url = $this->router->generate('security_login');

        return new RedirectResponse($url);
    }

    /**
     * @param Request $request
     * @param TokenInterface $token
     * @param string $providerKey The provider (i.e. firewall) key
     *
     * @return Response|null
     */
    public function onAuthenticationSuccess(Request $request, TokenInterface $token, $providerKey)
    {
        $url = $this->router->generate('homepage');

        return new RedirectResponse($url);
    }

    /**
     * @param Request $request
     * @param AuthenticationException $authException
     *
     * @return Response
     */
    public function start(Request $request, AuthenticationException $authException = null)
    {
        $url = $this->router->generate('security_login');

        return new RedirectResponse($url);
    }

    public function supportsRememberMe()
    {
        return true;
    }
}
```
And my FacebookController.php

```
<?php
namespace AppBundle\Controller\Connect;

use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
use Symfony\Component\HttpFoundation\Request;
use AppBundle\Entity\User;

/**
 * Class FacebookController
 * @Route("/connect/facebook")
 */
class FacebookController extends Controller
{
    /**
     * Link to this controller to start the "connect" process
     *
     * @Route("/", name="connect_facebook")
     */
    public function connectAction(Request $request)
    {
        // will redirect to Facebook!
        return $this->get('oauth2.registry')
            ->getClient('facebook') // key used in config.yml
            ->redirect();
            // I tried the following:
            // ->redirect([], ['redirect_params' => $request->headers->get('referer')]);
    }

    /**
     * After going to Facebook, you're redirected back here
     * because this is the "redirect_route" you configured
     * in config.yml
     *
     * @Route("/check", name="connect_facebook_check")
     */
    public function connectCheckAction(Request $request)
    {
    }
}
```

Is there a way to do this using redirect_params so the callback url keeps the user visited page? Right now I'm able to send additional parameters to facebook but it won't give it back to me in the callback url (connect_facebook_check). It seems like a basic functionality and I'm pretty sure it's already implemented but I can't get hold of it.

Or maybe I'm just misusing this code \o\

**Comment by weaverryan** [comment]
Hi there!

I may have a quick answer :). Have you looked st the TargetPathTrait in Symfony? Symfony automatically stores the last URL the visitor tried to access in the session. By using that trait in your authenticator, you can redirect to that url, if there is one. You can even manually set the URL the user should be redirected to if your requirements are a bit more custom.

Let me know if that helps!

**Comment by rainje** [comment]
Hi!

Thanks for your answer, I took a look at this trait but it seems it's only for symfony 3.1+ and I'm currently running 2.8.

So I added a listener to save last routes in sessions. Then when my route connect_facebook is accessed I save that last route in another session var so it's not overwritten by connect_facebook_check.

I may upgrade symfony to a newer version to use the built-in trait later ;)

Thanks for helping me get on a right track.
